In [148]:
import pandas as pd
import numpy as np

In [149]:
geo = pd.read_csv("DataSet/olist_geolocation_dataset.csv")
geo = geo.drop_duplicates(subset=['geolocation_zip_code_prefix'])

In [150]:
customers = pd.read_csv("DataSet/olist_customers_dataset.csv")
customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [151]:
sellers = pd.read_csv("DataSet/olist_sellers_dataset.csv")
sellers.head()

seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP

## Zip_Code

### Understanding the Zip_Code
Brinco De Ouro Da Princessa
Av. Imperatriz Dona Tereza Cristina, 87
Jardim Guarani, Campinas
SP, 13100-200, Brasile


    1 – the first character indicates the region, in our case São Paulo
    3 – the second character indicates the sub-region, Campinas
    1 – the third character indicates the sector, Campinas and surroundings
    0 – the fourth character indicates the sub-sector
    0 – the fifth character indicates the sub-sector divider
    200 – the last three characters indicate the delivery area, respectively the exact address

ref - https://www.europacco.com/en/find-zip/br?country=br

In [152]:
geo.head()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   
5                         1012       -23.547762       -46.635361   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP  
5        são paulo                SP

### Finding 1
Upon checking the first few rows, I see only four digit in zip_code column. I check to see the distribution of length of zip_code in dataset.

In [153]:
zip_code_length = geo["geolocation_zip_code_prefix"].astype(str).str.len()
zip_code_length.value_counts()

geolocation_zip_code_prefix
5    14661
4     4354
Name: count, dtype: int64

In [154]:
geo.dtypes

geolocation_zip_code_prefix      int64
geolocation_lat                float64
geolocation_lng                float64
geolocation_city                object
geolocation_state               object
dtype: object

`geolocation_zip_code_prefix` should be str type.
I checked the first row which has `geolocation_zip_code_prefix` 1037, by using (lat, lng) value in google map.
Turns out the zip_code value is missing the first 0.

In [155]:
def update_zip_code(df, col):
    df[col] = np.where(
        df[col].astype(str).str.len() == 4, 
        '0' + df[col].astype(str), 
        df[col].astype(str)
    )

update_zip_code(geo, "geolocation_zip_code_prefix")
geo.head()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                       01037       -23.545621       -46.639292   
1                       01046       -23.546081       -46.644820   
3                       01041       -23.544392       -46.639499   
4                       01035       -23.541578       -46.641607   
5                       01012       -23.547762       -46.635361   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP  
5        são paulo                SP

In [156]:
update_zip_code(customers, "customer_zip_code_prefix")
customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

  customer_zip_code_prefix          customer_city customer_state  
0                    14409                 franca             SP  
1                    09790  sao bernardo do campo             SP  
2                    01151              sao paulo             SP  
3                    08775        mogi das cruzes             SP  
4                    13056               campinas             SP

In [157]:
update_zip_code(sellers, "seller_zip_code_prefix")
sellers.head()

seller_id seller_zip_code_prefix        seller_city  \
0  3442f8959a84dea7ee197c632cb2df15                  13023           campinas   
1  d1b65fc7debc3361ea86b5f14c68d2e2                  13844         mogi guacu   
2  ce3ad9de960102d0677a81f5d0bb7b2d                  20031     rio de janeiro   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                  04195          sao paulo   
4  51a04a8a6bdcb23deccc82b0b80742cf                  12914  braganca paulista   

  seller_state  
0           SP  
1           SP  
2           RJ  
3           SP  
4           SP

In [158]:
geo.columns = ['zip_code', 'lat', 'lng', 'city', 'state']
customers.columns = ['customer_id', 'customer_unique_id', 'zip_code', 'city', 'state']
sellers.columns = ['seller_id', 'zip_code', 'city', 'state']

### Finding 2
When I tried to join sellers and geo, geo dataset have **missing values**.
I tried finding the (lat, lng) value on google map by using the nearest zip_code.
Turn out they are not exactly correct.
I'll filled the missing values by using the **nearest value**.

In [159]:
print(geo.shape)
geo = pd.merge(sellers[['zip_code']], geo, on = 'zip_code', how = 'outer')
geo.sort_values(by='zip_code', inplace=True)

# Fills all columns
geo = geo.ffill()
print(geo.shape)

(19015, 5)
(19871, 5)


In [160]:
geo = pd.merge(customers[['zip_code']], geo, on = 'zip_code', how = 'outer')
geo.sort_values(by='zip_code', inplace=True)

# Fills all columns
geo = geo.ffill()
print(geo.shape)

(120628, 5)


In [162]:
geo = geo.drop_duplicates(subset=['zip_code'])
print(geo.shape)

(19177, 5)


### Save Cleaned Data


In [163]:
geo.to_csv('geo.csv', index=False)
customers.to_csv('customers.csv', index=False)
sellers.to_csv('sellers.csv', index=False)